internal/parse/parser.go

```diff
@@ -141,13 +143,38 @@ func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
 		}
 
 		// If nothing above can be parsed we advance the parser.
-		p.pos++
+		p.advance()
 	}
 	// Add any remaining unparsed string input to the parser.
 	p.add(nil)
 	return &ParsedExpr{p.parts}, nil
 }
 
+// advance increments p.pos until we reach a space, a $, a " or a '. If we find
+// a space we advance to the rightmost adjacent space so the p.pos points to the
+// last space before the next non-space character.
+func (p *Parser) advance() {
+	noteableBytes := map[byte]bool{
+		'$':  true,
+		'"':  true,
+		'\'': true,
+		' ':  true,
+	}
+	// Advance the parser to the next char.
+	p.pos++
+	// Keep incrementing pos until we find a character of interest.
+	for p.pos < len(p.input) && !noteableBytes[p.input[p.pos]] {
+		p.pos++
+	}
+	// If we are on a space, advance to the rightmost in the sequence.
+	if p.peekByte(' ') {
+		for p.pos+1 < len(p.input) && p.input[p.pos+1] == ' ' {
```

Comment:
So, about the actual behavior: this skipping towards any "notable" char, such as `$`, even if it hasn't been preceded by something that is a real separator. For example, this would advance to the `$`  in `asd$asd`, even though this isn't valid syntax. Also, the opposite case is interesting: the current logic stops at spaces, even if the next character isn't something that could be relevant to the expression language. So, I guess the question is: what logic should we implement here?  This seems close, but doesn't seem to be quite there yet.

Reply:
As the [spec](https://docs.google.com/document/d/11WUBE5Ik5t113pWzMq78H8nXT_Rv_YIawDZpziSV0_k/edit#heading=h.2bm4jvbrvoef) stands, the part of `asd$asd` following the `$` is considered as an input expression (though not a well formed one). We don't require that there is a space before the `$`. The thinking being that a user may well write `p=$Person.name` or `($Person.name)`.

We could change the input parser so that we only recognise it as an input expression if `$` is followed by a capital letter and/or the type is qualified by a column (though this would be a separate PR).

We stop at spaces because we have to watch out for the columns which come before the expression e.g. `col1 AS &P.col1` or `(col1, col2) AS &P.*. We could skip to the next space that is followed by an alphanumeric char but since that includes nearly everything likely to follow a space it would not be much of a speed up.

An alternative might be to skip the next `&` and somehow look backwards for the columns, though this would require a rewrite of #23.

Reply:
For completeness so we can close this point, since we discussed this a while ago in our calls: the specification doesn't have a precise syntax definition for the syntax, and thus is prone to interpretation and discussion during implementation. With that said, assuming that "foo$asd" is a valid expression when the spec was talking about $Type.field seems like an unresonable stretch.